### PR TITLE
Typeset entire exercise again & use context to typeset from some children

### PR DESCRIPTION
**Pull request by jivey**
for: https://github.com/openstax/unified/issues/2443
Reverts the previous change to only typeset Question, so the entire exercise is again typeset as a single pass. Use context to call a typeset from child components if question feedback or detailed solutions change.

### CODE REVIEW

File: src/components/Exercise.tsx [hunk 1]
```diff
@@ -86,41 +95,49 @@ export const Exercise = ({
     }
   }, [scrollToQuestion, exercise]);
 
-  return (<StyledTaskStepCard
+  React.useEffect(() => {
+    typesetExercise();
+  }, [exercise]);
```
Comment (TomWoodward) [inline src/components/Exercise.tsx:100]: you could not have this here and just call the context from `Content`

File: src/hooks/useTypesetMath.ts [hunk 2]
```diff
@@ -0,0 +1,4 @@
+import React from "react";
+
+export const TypesetMathContext = React.createContext<{ typesetMath?: () => void }>({});
+export const useTypesetMath = () => React.useContext(TypesetMathContext);
```
Comment (TomWoodward) [inline src/hooks/useTypesetMath.ts:4]: if you initialize to a placeholder you can get rid of your null checks in the usage:

```suggestion
export const TypesetMathContext = React.createContext<() => void>(() => { throw new Error('context not initialized'))});
export const useTypesetMath = () => React.useContext(TypesetMathContext);
```

**Review by TomWoodward — CHANGES_REQUESTED** [review]

File: src/components/Exercise.tsx (hunk 1 above)
Comment (jivey) [reply src/components/Exercise.tsx:100]: Yeeaaaa I like that a lot better!

File: src/hooks/useTypesetMath.ts (hunk 2 above)
Comment (jivey) [reply src/hooks/useTypesetMath.ts]: I also like that this forces thinking about handling math content if `Content` gets used outside of the Exercise context.

File: src/components/Content.tsx [hunk 3]
```diff
@@ -15,6 +16,12 @@ export interface ContentProps<T extends ComponentType | undefined> {
 export const Content = (<T extends ComponentType | undefined>(
   {html, component, block = false, ...props}: ContentProps<T>
 ) => {
+  const typesetMath = useTypesetMath();
+
+  React.useEffect(() => {
+    typesetMath();
+  }, [html])
```
Comment (TomWoodward) [inline src/components/Content.tsx:23]: this seems like it should be complaining about the missing dependency in the array

File: src/components/Exercise.tsx [hunk 4]
```diff
@@ -78,6 +80,13 @@ export const Exercise = ({
 }: ExerciseWithStepDataProps | ExerciseWithQuestionStatesProps) => {
   const legacyStepRender = 'feedback_html' in step;
   const questionsRef = React.useRef<Array<HTMLDivElement>>([]);
+  const container = React.useRef<HTMLDivElement>(null);
+
+  const typesetExercise = () => {
+    if (container.current) {
+      typesetMath(container.current);
+    }
+  };
```
Comment (TomWoodward) [inline src/components/Exercise.tsx:89]: you may want to wrap this in a React.useCallback, especially if you need to add it to the useEffect dependencies

**Review by TomWoodward — APPROVED** [review]

**Pull request merged**